bluepill/src/BPPacker.m

```diff
@@ -86,5 +96,73 @@ @implementation BPPacker
     return bundles;
 }
 
++ (NSMutableArray<BPXCTestFile *> *)packTestsByTime:(NSArray<BPXCTestFile *> *)xcTestFiles
+                                      configuration:(BPConfiguration *)config
+                                           andError:(NSError **)errPtr {
+    [BPUtils printInfo:INFO withString:@"Packing based on individual test execution times in file path: %@", config.testTimeEstimatesJsonFile];
+    if (xcTestFiles.count == 0) {
+        BP_SET_ERROR(errPtr, @"Found no XCTest files.\n"
+                     "Perhaps you forgot to 'build-for-testing'? (Cmd + Shift + U) in Xcode.");
+        return NULL;
+    }
+
+    // load the config file
+    NSDictionary *testTimes = [self loadConfigFile:config.testTimeEstimatesJsonFile withError:errPtr];
+    if ((errPtr && *errPtr) || !testTimes) {
+        [BPUtils printInfo:ERROR withString:@"%@", [*errPtr localizedDescription]];
+        return NULL;
+    }
+
+    NSMutableDictionary *estimatedTimesByTestFilePath = [[NSMutableDictionary alloc] init];
+    NSArray *testCasesToRun = config.testCasesToRun;
+
+    double totalTime = 0.0;
+    for (BPXCTestFile *xctFile in xcTestFiles) {
+        NSMutableSet *bundleTestsToRun = [[NSMutableSet alloc] initWithArray:[xctFile allTestCases]];
+        if (testCasesToRun) {
+            [bundleTestsToRun intersectSet:[[NSSet alloc] initWithArray:testCasesToRun]];
+        }
+        if (config.testCasesToSkip && [config.testCasesToSkip count] > 0) {
+            [bundleTestsToRun minusSet:[[NSSet alloc] initWithArray:config.testCasesToSkip]];
+        }
+        if (bundleTestsToRun.count > 0) {
+            double __block testBundleExecutionTime = 0.0;
+            [bundleTestsToRun enumerateObjectsUsingBlock:^(id _Nonnull test, BOOL * _Nonnull stop) {
+                // TODO: Assign a sensible default if the estimate is not given
+                if ([testTimes objectForKey:test]) {
+                    testBundleExecutionTime += [[testTimes objectForKey:test] doubleValue];
+                }
+            }];
+            estimatedTimesByTestFilePath[xctFile.testBundlePath] = [NSNumber numberWithDouble:testBundleExecutionTime];
+            totalTime += testBundleExecutionTime;
+        }
+    }
+    assert([estimatedTimesByTestFilePath count] == [xcTestFiles count]);
+
+    NSMutableArray<BPXCTestFile *> *bundles = [[NSMutableArray alloc] init];
+    for (BPXCTestFile *xctFile in xcTestFiles) {
+        BPXCTestFile *bundle = [xctFile copy];
+        bundle.skipTestIdentifiers = config.testCasesToSkip;
+        [bundle setEstimatedExecutionTime:estimatedTimesByTestFilePath[xctFile.testBundlePath]];
+        [bundles addObject:bundle];
+    }
+    // Sort bundles by execution times
```

Comment:
Maybe mention that it's a reverse ordering here...

Reply:
Done.